streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -136,158 +233,149 @@ public File baseDir() {
+                stateRestoreCallback,
+                converterForStore(store));
+            stores.put(storeName, storeMetadata);
+
+            changelogReader.register(storePartition, this);
+        } else {
+            stores.put(storeName, new StateStoreMetadata(store));
         }
 
-        registeredStores.put(storeName, Optional.of(store));
+        log.debug("Registered state store {} to its state manager", storeName);
     }
 
     @Override
-    public void reinitializeStateStoresForPartitions(final Collection<TopicPartition> partitions,
-                                                     final InternalProcessorContext processorContext) {
-        StateManagerUtil.reinitializeStateStoresForPartitions(log,
-                                                              eosEnabled,
-                                                              baseDir,
-                                                              registeredStores,
-                                                              storeToChangelogTopic,
-                                                              partitions,
-                                                              processorContext,
-                                                              checkpointFile,
-                                                              checkpointFileCache
-        );
+    public StateStore getStore(final String name) {
+        if (stores.containsKey(name)) {
+            return stores.get(name).stateStore;
+        } else {
+            return null;
+        }
     }
 
-    void clearCheckpoints() throws IOException {
-        if (checkpointFile != null) {
-            checkpointFile.delete();
-            checkpointFile = null;
+    Collection<TopicPartition> changelogPartitions() {
+        return changelogOffsets().keySet();
+    }
 
-            checkpointFileCache.clear();
+    @Override
+    public Map<TopicPartition, Long> changelogOffsets() {
+        // return the current offsets for those logged stores
+        final Map<TopicPartition, Long> changelogOffsets = new HashMap<>();
+        for (final StateStoreMetadata storeMetadata : stores.values()) {
+            if (storeMetadata.changelogPartition != null) {
+                // for changelog whose offset is unknown, use 0L indicating earliest offset
+                // otherwise return the current offset + 1 as the next offset to fetch
+                changelogOffsets.put(
+                    storeMetadata.changelogPartition,
+                    storeMetadata.offset == null ? 0L : storeMetadata.offset + 1L);
+            }
         }
+        return changelogOffsets;
     }
 
-    @Override
-    public Map<TopicPartition, Long> checkpointed() {
-        updateCheckpointFileCache(emptyMap());
-        final Map<TopicPartition, Long> partitionsAndOffsets = new HashMap<>();
+    TaskId taskId() {
+        return taskId;
+    }
+
+    // used by the changelog reader only
```

Comment:
these kinds of comments paradoxically lead to unmaintainability. Either a method is part of the public contract or not. If it is, then this comment would become out of date, if not, then the changelog reader shouldn't be using it. The recommendation is simply to delete the comment (and similar ones). This even applies to "for testing" comments. I have found several methods in use in this code base that were commented "visible for testing". Even for tests, either move both the class and the test into an isolated package and use package-private, refactor the test, or remove the comment. Also, if the changelog reader really needs four "holes" poked into this class, then we should reconsider the relationship between the state manager and the changelog reader.